mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java

```diff
@@ -0,0 +1,444 @@
+      FileIO io, boolean throwOnFailure) {
+    JobConf conf = jobContext.getJobConf();
+    // If there are reducers, then every reducer will generate a result file.
+    // If this is a map only task, then every mapper will generate a result file.
+    int expectedFiles = conf.getNumReduceTasks() > 0 ? conf.getNumReduceTasks() : conf.getNumMapTasks();
+
+    Collection<DataFile> dataFiles = new ConcurrentLinkedQueue<>();
+
+    // Reading the committed files. The assumption here is that the taskIds are generated in sequential order
+    // starting from 0.
+    Tasks.range(expectedFiles)
+        .throwFailureWhenFinished(throwOnFailure)
+        .executeWith(executor)
+        .retry(3)
+        .run(taskId -> {
+          String taskFileName = generateFileForCommitLocation(location, conf, jobContext.getJobID(), taskId);
+          dataFiles.addAll(Arrays.asList(readFileForCommit(taskFileName, io)));
+        });
+
+    return dataFiles;
+  }
+
+  /**
+   * Generates the job temp location based on the job configuration.
+   * Currently it uses TABLE_LOCATION/temp/QUERY_ID-jobId.
+   * @param location The location of the table
+   * @param conf The job's configuration
+   * @param jobId The JobID for the task
+   * @return The file to store the results
+   */
+  @VisibleForTesting
+  static String generateJobLocation(String location, Configuration conf, JobID jobId) {
+    String queryId = conf.get(HiveConf.ConfVars.HIVEQUERYID.varname);
+    return location + "/temp/" + queryId + "-" + jobId;
+  }
+
+  /**
+   * Generates file location based on the task configuration and a specific task id.
+   * This file will be used to store the data required to generate the Iceberg commit.
+   * Currently it uses TABLE_LOCATION/temp/QUERY_ID-jobId/task-[0..numTasks).forCommit.
+   * @param location The location of the table
+   * @param conf The job's configuration
+   * @param jobId The jobId for the task
+   * @param taskId The taskId for the commit file
+   * @return The file to store the results
+   */
+  private static String generateFileForCommitLocation(String location, Configuration conf, JobID jobId, int taskId) {
+    return generateJobLocation(location, conf, jobId) + "/task-" + taskId + FOR_COMMIT_EXTENSION;
+  }
+
+  private static void createFileForCommit(DataFile[] closedFiles, String location, FileIO io)
+      throws IOException {
+
+    OutputFile fileForCommit = io.newOutputFile(location);
+    try (ObjectOutputStream oos = new ObjectOutputStream(fileForCommit.createOrOverwrite())) {
+      oos.writeObject(closedFiles);
+    }
+    LOG.debug("Iceberg committed file is created {}", fileForCommit);
+  }
+
+  private static DataFile[] readFileForCommit(String fileForCommitLocation, FileIO io) {
+    try (ObjectInputStream ois = new ObjectInputStream(io.newInputFile(fileForCommitLocation).newStream())) {
+      return (DataFile[]) ois.readObject();
+    } catch (ClassNotFoundException | IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)